File: postm/controllers/posts.py
from flask import request

from postm.services.posts import PostsService
from postm.repositories.posts import PostPage

from os import getenv

class PostsController(object):
    def __init__(self) -> None:
        self.service = PostsService()

    def create(self) -> tuple[dict, int]:

        data = request.form

        title: str = data.get("title", "")
        description: str = data.get("description", "")
        image = request.files.get("image", None)

        try:

            post = self.service.create(
                title = title,
                description = description,
                image = image
            )

            if post.image:
                post.image = f"{ getenv("URL") }/api/images/{ post.image }"

            return post.toJson(), 200

        except Exception as error:
            return {
                "error": str(error)
            }, 400
        
    def findAll(self) -> tuple[dict, int]:
        try: 
            posts = self.service.findAll()

            postsFormated = [post.toJson() for post in posts]

            return { "posts": postsFormated }, 200
        
        except: 
            return {
                "error": "Something unexpected happened"
            }, 500
    
    def findById(self, id: str) -> tuple[dict, int]:
        try:
            post = self.service.findById(id)

            return post.toJson(), 200
        
        except Exception as error:
            return {
                "error": str(error)
            }, 400

    def delete(self, id: str) -> tuple[dict, int]:
        try:
            if not self.service.delete(id): 
                return {
                    "message": f"Unable to remove post { id }"
                }, 400
            
            return {
                "message": f"post { id } removed"
            }, 200
    
        except Exception as error:
            return {
                "error": str(error)
            }, 400

    def update(self, id: str) -> tuple[dict, int]:

        data = request.form

        title: str = data.get("title", "")
        description: str = data.get("description", "")
        image = request.files.get("image", None)

        try:
            isUpdate = self.service.update(
                id = id,
                title = title,
                description = description,
                image = image
            )

            if not isUpdate: 
                return {
                    "message": f"Unable to update post { id }"
                }, 400
            
            return {
                "message": f"post { id } updated"
            }, 200

        except Exception as error:
            return {
                "error": str(error)
            }, 400
        
    def createNextUrlPage(self, page: PostPage) -> str:
        query = f"index={ page.index + 1 }&size={ page.size }"

        return f"{ getenv("URL") }/api/posts/all/page?{ query }"
    
    def createPreviousUrlPage(self, page: PostPage) -> str:
        if page.index - 1 < 0:
            return None 

        query = f"index={ page.index - 1 }&size={ page.size }"

        return f"{ getenv("URL") }/api/posts/all/page?{ query }"
        
    def findAllPaged(self) -> tuple[dict, int]:
        index = request.args.get("index", 0)
        size = request.args.get("size", 10)

        try: 
            index = int(index)
            size = int(size)

            postPage = self.service.findAllPaged(
                index = int(index),
                size = int(size)
            )

            postFormated = [ p.toJson() for p in postPage.posts ]

            nextUrl = self.createNextUrlPage(
                page = postPage
            )

            if not postFormated:
                nextUrl = None

            previousUrl = self.createPreviousUrlPage(
                page = postPage
            )

            return {
                "posts": postFormated,
                "next": nextUrl,
                "previous": previousUrl
            }, 200
        
        except Exception as error:
            return {
                "error": str(error)
            }, 400
        

File: src/controllers/posts.ts
import { Request, Response } from "express";
import Service from "../services/posts";
import { z } from "zod";
import env from "../env";

const createSchema = z.object({
    title: z.string({ message: "title required" }),
    description: z.string({ message: "title required" })
});

const idSchema = z.object({
    id: z.string({ message: "id required "})
});

const pageSchema = z.object({
    size: z.coerce.number().default(10),
    index: z.coerce.number().default(0)

});

const updateSchema = z.object({
    title: z.string({ message: "title required" }),
    description: z.string({ message: "title required" })
});

export default class Controller {

    private service: Service;

    public constructor() {
        this.service = new Service();
        this.create = this.create.bind(this);
        this.all = this.all.bind(this);
        this.find = this.find.bind(this);
        this.delete = this.delete.bind(this);
        this.allPaged = this.allPaged.bind(this);
        this.update = this.update.bind(this);
    }

    public async update(request: Request, response: Response): Promise<void> {
        try {
            const { id } = idSchema.parse(request.params);
            const { 
                title, 
                description 
            } = updateSchema.parse(request.body);

            let image = null;

            if (request.file) {
                image = request.file.filename;
            }

            const post = await this.service.update(
                id,
                title,
                description,
                image
            );

            response.status(200).json(post);

        } catch(error: any) {

            response.status(500).json({
                message: error.message
            });

        }
    }

    public async allPaged(request: Request, response: Response): Promise<void> {
        try {
            const { index, size } = pageSchema.parse(request.query);

            const hostUrl = `${ env.SERVER_HOST }:${ env.SERVER_PORT }`; 

            const page = await this.service.allPaged(index, size);

            for(const post of page.posts) {
                post.image = `${ hostUrl }/api/images/${ post.image }`;
            }

            response.status(200).json({
                posts: page.posts,
                next: `${ hostUrl }/posts/all/page?size=${ size }&index=${ index + 1}`,
                previous: `${ hostUrl }/posts/all/page?size=${ size }&index=${ index - 1}`
            });

        } catch(error: any) {

            response.status(500).json({
                message: error.message
            });

        }
    }

    public async delete(request: Request, response: Response): Promise<void> {
        try {
            const { id } = idSchema.parse(request.params);

            if (!await this.service.delete(id)) {
                response.status(400).json({
                    message: `Unable to remove post ${ id }`
                })                    
                return;
            }
            
            response.status(200).json({
                message: `post ${ id } removed`
            });

        } catch(error: any) {

            if (error instanceof z.ZodError) {
                response.status(400).json({
                    message: error.issues[0].message
                });
                return;
            }

            response.status(500).json({
                message: error.message
            });

        }

    }

    public async find(request: Request, response: Response): Promise<void> {
        try {
            const { id } = idSchema.parse(request.params);

            const post = await this.service.findById(id);

            if (!post) {
                response.status(400).json({
                    message: `post ${ id } not exists`
                })                    
                return;
            }
            
            response.status(200).json(post);

        } catch(error: any) {

            if (error instanceof z.ZodError) {
                response.status(400).json({
                    message: error.issues[0].message
                });
                return;
            }

            response.status(500).json({
                message: error.message
            });

        }

    }

    public async create(request: Request, response: Response): Promise<void> {
        try {
            const { 
                title, 
                description 
            } = createSchema.parse(request.body);

            let image = null;

            if (request.file) {
                image = request.file.filename;
            }

            const post = await this.service.create(
                title, 
                description, 
                image
            );

            const hostUrl = `${ env.SERVER_HOST }:${ env.SERVER_PORT }`; 

            post.image = `${ hostUrl}/api/images/${ post.image }`;

            response.status(200).json(post);

        } catch(error: any) {

            if (error instanceof z.ZodError) {
                response.status(400).json({
                    message: error.issues[0].message
                });
                return;
            }

            response.status(500).json({
                message: error.message
            });

        }
    }

    public async all(request: Request, response: Response): Promise<void> {
        try {

            const hostUrl = `${ env.SERVER_HOST }:${ env.SERVER_PORT }`; 

            const posts = await this.service.all();

            for(const post of posts) {
                post.image = `${ hostUrl}/api/images/${ post.image }`;
            }

            response.status(200).json({ posts });

        } catch(error: any) {

            response.status(500).json({
                message: error.message
            });

        }
    }

}